manager/shard_strategy.go

```diff
@@ -0,0 +1,191 @@
+}
+
+func NewShardStrategy(ctx context.Context, shardConfig config.ShardConfig) (ShardStrategy, error) {
+	switch shardConfig.Type {
+	case config.HashShardType:
+		if shardConfig.PodCount <= 0 {
+			return nil, fmt.Errorf("configured PodCount (%d) must be greater than zero", shardConfig.PodCount)
+		} else if shardConfig.PodCount > v1alpha1.ShardKeyspaceSize {
+			return nil, fmt.Errorf("configured PodCount (%d) is larger than available keyspace size (%d)", shardConfig.PodCount, v1alpha1.ShardKeyspaceSize)
+		}
+
+		return &HashShardStrategy{
+			enableUncoveredReplica: shardConfig.EnableUncoveredReplica,
+			podCount:               shardConfig.PodCount,
+		}, nil
+	case config.ProjectShardType, config.DomainShardType:
+		replicas := make([][]string, 0)
+		for _, replica := range shardConfig.Replicas {
+			if len(replica.Entities) == 0 {
+				return nil, fmt.Errorf("unable to create replica with 0 configured entity(ies)")
+			}
+
+			replicas = append(replicas, replica.Entities)
+		}
+
+		var envType environmentType
+		switch shardConfig.Type {
+		case config.ProjectShardType:
+			envType = project
+		case config.DomainShardType:
+			envType = domain
+		}
+
+		return &EnvironmentShardStrategy{
+			enableUncoveredReplica: shardConfig.EnableUncoveredReplica,
+			envType:                envType,
+			replicas:               replicas,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("shard strategy '%s' does not exist", shardConfig.Type)
+}
+
+// HashShardStrategy evenly assigns disjoint keyspace responsibilities over a collection of pods.
+// All FlyteWorkflows are assigned a shard-key using a hash of their executionID and are then
+// processed by the FlytePropeller instance responsible for that keyspace range.
+type HashShardStrategy struct {
+	enableUncoveredReplica bool
+	podCount               int
+}
+
+func (h *HashShardStrategy) GetPodCount() int {
+	if h.enableUncoveredReplica {
+		return h.podCount + 1
+	}
+
+	return h.podCount
+}
+
+func (h *HashShardStrategy) UpdatePodSpec(pod *v1.PodSpec, podIndex int) error {
+	container, err := getFlytePropellerContainer(pod)
+	if err != nil {
+		return err
+	}
+
+	if podIndex >= 0 && podIndex < h.podCount {
+		startKey, endKey := computeKeyRange(v1alpha1.ShardKeyspaceSize, h.podCount, podIndex)
+		for i := startKey; i < endKey; i++ {
+			container.Args = append(container.Args, "--propeller.include-shard-label", fmt.Sprintf("%d", i))
+		}
+	} else if h.enableUncoveredReplica && podIndex == h.podCount {
+		for i := 0; i < v1alpha1.ShardKeyspaceSize; i++ {
+			container.Args = append(container.Args, "--propeller.exclude-shard-label", fmt.Sprintf("%d", i))
+		}
+	} else {
+		return fmt.Errorf("invalid podIndex '%d' out of range [0,%d)", podIndex, h.GetPodCount())
+	}
+
+	return nil
+}
+
+// Computes a [startKey, endKey) pair denoting the key responsibilities for the provided pod index
+// given the keyspaceSize and podCount parameters
+func computeKeyRange(keyspaceSize, podCount, podIndex int) (int, int) {
```

Comment:
I think this should be an exported method or util method and have tests on it

Reply:
I exported it. The testing should be very thorough as is, we're testing for increasing podCounts up to a max test keyspaceSize of 32. Each test ensures that the lowest and highest number of keys assigned to a specific pod doesn't vary by more than 1.

I'll add a note to refactor this out to a util method when we refactor the project structure.